### PR TITLE
[TCP] templated unary ops lowering

### PR DESCRIPTION
**Pull request by zezhang**
- Use a standard template to lower unary ops that share the same lowering process.
      - Only Ranked Tensor types are supported for now in template
      - Only floating-point datatype are supported for now in template
- Abs Op uses a specialized lowering method to support both integer (math.absi)and fp (math.absf) ops

To test:

`./build/bin/llvm-lit externals/llvm-external-projects/torch-mlir-dialects/test/Conversion/TcpToLinalg/ -v`
`./build/bin/llvm-lit externals/llvm-external-projects/torch-mlir-dialects/test/Dialect/Tcp/ -v`
`./build/bin/llvm-lit test/Conversion/TorchToTcp/ -v`
`python -m e2e_testing.main --config=tcp -v -s`

### CODE REVIEW

File: externals/llvm-external-projects/torch-mlir-dialects/test/Conversion/TcpToLinalg/unary.mlir [hunk 1]
```diff
@@ -166,3 +166,138 @@ func.func @floor(%arg0 : tensor<?x?xf32>) -> tensor<?x?xf32> {
...
+// CHECK:         ^bb0(%[[BBARG0:.*]]: f32, %{{.*}}: f32):
+// CHECK:           %[[CEIL:.*]] = math.absf %[[BBARG0]] : f32
+// CHECK:           linalg.yield %[[CEIL]] : f32
+// CHECK:         } -> tensor<?x?xf32>
+// CHECK:         return %[[GENERIC]] : tensor<?x?xf32>
+// CHECK:       }
+func.func @abs(%arg0 : tensor<?x?xf32>) -> tensor<?x?xf32> {
+  %0 = tcp.abs %arg0 : tensor<?x?xf32> -> tensor<?x?xf32>
+  return %0 : tensor<?x?xf32>
+}
```
Comment (sjain-stanford) [inline externals/llvm-external-projects/torch-mlir-dialects/test/Conversion/TcpToLinalg/unary.mlir:249]: We'd need a similar test for the integer variant.

File: externals/llvm-external-projects/torch-mlir-dialects/test/Dialect/Tcp/unary.mlir [hunk 2]
```diff
@@ -106,3 +106,58 @@ func.func @test_floor_f32(%arg0 : tensor<?x?xf32>) -> tensor<?x?xf32> {
...
+// -----
+
+// CHECK-LABEL: func.func @test_abs_f32(
+// CHECK-SAME:               %[[ARG:.*]]: tensor<?x?xf32>) -> tensor<?x?xf32>
+// CHECK:         %[[ABS:.*]] = tcp.abs %[[ARG]] : tensor<?x?xf32> -> tensor<?x?xf32>
+// CHECK:         return %[[ABS]] : tensor<?x?xf32>
+func.func @test_abs_f32(%arg0 : tensor<?x?xf32>) -> tensor<?x?xf32> {
+  %0 = tcp.abs %arg0 : tensor<?x?xf32> -> tensor<?x?xf32>
+  return %0 : tensor<?x?xf32>
+}
```
Comment (sjain-stanford) [inline externals/llvm-external-projects/torch-mlir-dialects/test/Dialect/Tcp/unary.mlir:141]: For completeness, please add a test for `tcp.abs` operating on integer types.

File: test/Conversion/TorchToTcp/elementwise.mlir [hunk 3]
```diff
@@ -146,6 +146,19 @@ func.func @torch.aten.clamp(%arg0: !torch.vtensor<[?,?],f32>) -> !torch.vtensor<
 
 // -----
 
+// CHECK-LABEL:  func.func @torch.aten.tanh(
+// CHECK-SAME:         %[[ARG0:.*]]: !torch.vtensor<[?,?],f32>) -> !torch.vtensor<[?,?],f32> {
+// CHECK:         %[[T0:.*]] = torch_c.to_builtin_tensor %[[ARG0]] : !torch.vtensor<[?,?],f32> -> tensor<?x?xf32>
+// CHECK:         %[[T1:.*]] = tcp.tanh %[[T0]] : tensor<?x?xf32> -> tensor<?x?xf32>
+// CHECK:         %[[T2:.*]] = torch_c.from_builtin_tensor %[[T1]] : tensor<?x?xf32> -> !torch.vtensor<[?,?],f32>
+// CHECK:         return %[[T2]] : !torch.vtensor<[?,?],f32>
+func.func @torch.aten.tanh(%arg0: !torch.vtensor<[?,?],f32>) -> !torch.vtensor<[?,?],f32> {
```
Comment (sjain-stanford) [inline test/Conversion/TorchToTcp/elementwise.mlir:155]: Thanks for adding this, could've been missed previously.

File: lib/Conversion/TorchToTcp/Elementwise.cpp [hunk 4]
```diff
@@ -440,23 +335,64 @@ class ConvertAtenReluOp : public OpConversionPattern<AtenReluOp> {
...
+      return rewriter.notifyMatchFailure(
+          op, "Abs input tensor must have integer or floating-point datatype");
+
+    rewriter.replaceOpWithNewOp<tcp::AbsOp>(op, inputType, input);
+    return success();
+  }
+};
+
+template <typename AtenOpT, typename TcpOpT>
+class ConvertAtenTemplatedUnaryOp : public OpConversionPattern<AtenOpT> {
```
Comment (sjain-stanford) [inline lib/Conversion/TorchToTcp/Elementwise.cpp:361]: nit: consider dropping `Templated` from the name (as it is self-explanatory)

**Review by sjain-stanford — COMMENTED** [review]
Thanks for adding more unary ops and templatizing the TorchToTcp conversion. Requested adding missing tests and minor nits. Otherwise LGTM.

File: lib/Conversion/TorchToTcp/Elementwise.cpp (hunk 4 above)
Comment (navahgar) [reply lib/Conversion/TorchToTcp/Elementwise.cpp:361]: +1 to this

**Review by navahgar — COMMENTED** [review]
Mostly LGTM. Some minor nits.

**Review by sjain-stanford — APPROVED** [review]
LGTM

**Pull request merged**